### ISSUE TITLE
systemd-networkd: no ipv6 address from ISP with >=systemd-networkd-251 via dhcpv6

### ISSUE THREAD

**Issue by borisdigital** [systemd/systemd#24292]
### systemd version the issue has been seen with

sys-apps/systemd-251.x

### Used distribution

Gentoo

### Linux kernel version used

5.15.59-gentoo-dist

### CPU architectures issue was seen on

x86_64

### Expected behaviour you didn't see

My router (apu4d4 with gentoo linux) should get an ipv6 address (via dhcpv6) from my ISP (Deutsche Glasfaser). This works fine when using systemd-250.7 (and lower).

### Unexpected behaviour you saw

My router gets an ipv4 address, but no ipv6 address. Additionally the ipv6 prefix isn't retrieved as well.

### Steps to reproduce the problem

Since systemd-251 i'm not able to get an ipv6 address via dhcpv6 from my ISP. This has been tested with archlinux (tried on july somehow, systemd-251.x, dunno anymore) and (since about one month) with gentoo linux. To be honest, this bug report (https://github.com/systemd/systemd/issues/23546) did really sound like _my_ problem, but this is _not_ an ppp interface, but a very _normal_ (intel) ethernet device (with 1G speed) connected to an fiber2lan bridge (device whatever) and i did test git master branch from 2022-08-11, which should(?) include the fix(es) from https://github.com/systemd/systemd/pull/24002.

Systemd versions tried:
* systemd-251.2
 * systemd-251.4
 * systemd-9999 (git master branch from 2022-08-11)

upstream config (/etc/systemd/network/enp4s0.network):
[enp4s0.network.txt](https://github.com/systemd/systemd/files/9312237/enp4s0.network.txt)

logs from systemd-journald (with Environment=SYSTEMD_LOG_LEVEL=debug):
[systemd-networkd-9999.log](https://github.com/systemd/systemd/files/9312265/systemd-networkd-9999.log)

Please tell me, what you need from me (infos, logs, etc.), you'll get it. Bisecting will be a bit more complicated (or maybe not possible), since this is a somehow slow apu (https://www.apu-board.de/#apu4) and, more important, a **productive** router.

PS: This config works fine with 250.x (ipv6 address + ipv6 pd)

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
...
Aug 11 18:41:57 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Sent Request
Aug 11 18:41:57 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Next retransmission in 55s
Aug 11 18:41:57 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Received Reply message with non-zero status: Binding unavailable
Aug 11 18:41:57 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Failed to process received reply message, ignoring: Invalid argument
Aug 11 18:42:52 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Sent Request
Aug 11 18:42:52 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Next retransmission in 1min 49s
Aug 11 18:42:52 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Received Reply message with non-zero status: Binding unavailable
Aug 11 18:42:52 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Failed to process received reply message, ignoring: Invalid argument
Aug 11 18:44:41 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Sent Request
Aug 11 18:44:41 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Next retransmission in 1min 56s
Aug 11 18:44:41 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Received Reply message with non-zero status: Binding unavailable
Aug 11 18:44:41 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Failed to process received reply message, ignoring: Invalid argument
Aug 11 18:46:37 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Sent Request
Aug 11 18:46:37 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Next retransmission in 1min 54s
Aug 11 18:46:37 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Received Reply message with non-zero status: Binding unavailable
Aug 11 18:46:37 mantenna systemd-networkd[1412520]: enp4s0: DHCPv6 client: Failed to process received reply message, ignoring: Invalid argument
...
```

**Comment by yuwata** [comment]
Could you provide dumped packets? Both failed (i.e. v251) and successful cases (i.e. v250) please.

**Comment by yuwata** [comment]
BTW, I briefly checked [RFC8415](https://www.rfc-editor.org/rfc/rfc8415.html), but there seems nothing about NoBinding error in reply for DHCP request. Also, the server returns only error code, but no description about the error... Hm...

**Comment by borisdigital** [comment]
@yuwata  
Here they are. I did capture these packets with the following command:
```tcpdump -vvvv -X -ttt -n -i enp4s0 "ip6 and udp and (port 546 or port 547)"```  
If you need something else, please give me a hint what magical tcpdump options you need.  
  
[tcpdump-dhcpv6-v250.7.log](https://github.com/systemd/systemd/files/9331677/tcpdump-dhcpv6-v250.7.log) (dhcpv6/ipv6 working)
[tcpdump-dhcpv6-v251.4.log](https://github.com/systemd/systemd/files/9331678/tcpdump-dhcpv6-v251.4.log) (dhcpv6/ipv6 not working)

**Comment by yuwata** [comment]
@borisdigital Thank you!
Could you test PR #24299, especially its second commit?

**Comment by borisdigital** [comment]
@yuwata  
This seems to work for me. I recompiled system-251.4 with both of your patches from https://github.com/systemd/systemd/pull/24299 and ipv6 addresses prefixes are back again.
  
Thanks a lot for your time and efforts! So, works fine for me, closing ;)

**Comment by yuwata** [comment]
@borisdigital Thank you for testing the PR!

Tentatively reopen this page. This will be closed automatically when #24299 is merged.
